koin-projects/koin-core/src/main/kotlin/org/koin/core/registry/BeanRegistry.kt

```diff
@@ -123,7 +123,11 @@ class BeanRegistry {
         val secondaryTypeDefinitions: ArrayList<BeanDefinition<*>> = definitionsSecondaryTypes[type]
                 ?: createSecondaryType(type)
 
-        secondaryTypeDefinitions.add(definition)
+        if (definition in secondaryTypeDefinitions){
+            secondaryTypeDefinitions[secondaryTypeDefinitions.indexOf(definition)] = definition
```

Comment:
not necessary to add the definition if already in secondary types?

Reply:
If the definition is already in the secondaryTypeDefinitions, we want to replace it.
To understand why this is needed, have a look at this test:

```
@Test
    fun `can declare and override a single with secondary type on the fly`() {

        val koin = koinApplication {
            printLogger()
            modules(emptyList())
        }.koin

        val a = Simple.Component1()
        val b = Simple.Component1()

        koin.declare(a, secondaryTypes = listOf(Simple.ComponentInterface1::class))
        koin.declare(b, secondaryTypes = listOf(Simple.ComponentInterface1::class), override = true)

        assertEquals(b, koin.get<Simple.Component1>())
        assertEquals(b, koin.get<Simple.ComponentInterface1>())
    }
```

The alternative would be to convert `secondaryTypeDefinitions` to be a `Map` rather than an `ArrayList` so definitions can't be added twice.

Reply:
yeah, or change it for a Set instead of a map

Reply:
> yeah, or change it for a Set instead of a map

Also with `Set` we would need to do something like:
```
if (definition in secondaryTypeDefinitions){
     secondaryTypeDefinitions.remove(definition)
}
secondaryTypeDefinitions.add(definition)
```
I don't know if that is better 🤔

Reply:
with a set, we would replace it directly with `add()`, no?

Reply:
Nope. The `add()` function will not replace if the collection already contains the item we're trying to add. From the MutableCollections.kt KDoc:

```kotlin
    /**
     * Adds the specified element to the collection.
     *
     * @return `true` if the element has been added, `false` if the collection does not support duplicates
     * and the element is already contained in the collection.
     */
    public fun add(element: E): Boolean

Reply:
ok :) `definitionsSecondaryTypes` has to stay as a map. Let's keep to your proposal